File: legacy/tools/livecheck.py
#!/usr/bin/env python3
"""Fields declared in a config table must be read by the app, not only by tests.

STABILITY_RULES carried `greenPerDay` and `amberPerDay` on eight parameters.
Nothing in the app ever read them — drift is graded by CONSISTENCY_RULES, on
the spread across a window. The only readers in the repository were the tests.

That is worse than dead code. A stage 3 audit found that alkalinity "graded
steady at 1.4 dKH a week, three times the published limit", corrected it, and
verified the correction by reading the same unused constant. A fix and a test
agreeing with each other about a number the app never consults, reported as one
of the most significant findings of that stage.

deadcode.py cannot see this: the constant IS read, just not by anything that
runs. This checks the source file alone.
"""
import re
import sys

path = sys.argv[1] if len(sys.argv) > 1 else "src/reef-console.jsx"
src = open(path, encoding="utf-8").read()

# Config tables worth policing: a flat object of per-parameter settings.
TABLES = ["STABILITY_RULES", "CONSISTENCY_RULES", "SAFE_BOUNDS", "KIT_PRECISION",
          "CORRECTION_MAX_RATE", "DOSE_ADVICE_RULES", "DOSE_DRIFT_TRIGGER"]

problems = []
for table in TABLES:
    m = re.search(r"const " + table + r"\s*=\s*\{", src)
    if not m:
        continue
    start = src.index("{", m.start())
    depth, k = 0, start
    while k < len(src):
        if src[k] == "{":
            depth += 1
        elif src[k] == "}":
            depth -= 1
            if depth == 0:
                break
        k += 1
    body = src[start:k]
    # Field names used inside the table's entries.
    fields = set(re.findall(r"[\{,]\s*(\w+):", body))
    outside = src[:m.start()] + src[k:]
    for field in sorted(fields):
        if field in ("why", "label", "unit", "mode", "displayPer"):
            continue          # descriptive, not consulted
        # read as .field or ["field"] anywhere else in the source
        if re.search(r"\.\s*" + field + r"\b", outside) or f'"{field}"' in outside or f"'{field}'" in outside:
            continue
        problems.append((table, field))

if problems:
    print("FAIL config fields the app never reads:")
    for table, field in problems:
        print(f"    {table}.{field}")
    print("    (delete them, or wire them up — a field only tests read is worse than dead)")
    sys.exit(1)

print(f"OK   every config field is read by the app ({len(TABLES)} tables checked)")

# ---------------------------------------------------------------------------
# Fields an engine SETS on its result that nothing ever reads.
#
# `out.previous` was set by assessCalcium alone — assessAlkalinity and
# assessMagnesium never set it — and read by nothing anywhere. A field one copy
# of three grew, that no consumer asked for. The config-table check above could
# not see it: it inspects declarations, and this is an assignment.

ENGINES = ["assessAlkalinity", "assessCalcium", "assessMagnesium"]
set_fields = {}
for engine in ENGINES:
    i = src.find("function " + engine)
    if i < 0:
        continue
    j = src.index(") {", i) + 2
    depth, k = 0, j
    while k < len(src):
        if src[k] == "{":
            depth += 1
        elif src[k] == "}":
            depth -= 1
            if depth == 0:
                break
        k += 1
    body = src[j:k]
    for field in set(re.findall(r"\bout\.(\w+)\s*=(?![=])", body)):
        set_fields.setdefault(field, set()).add(engine)

# Where a result field could be read: anywhere outside the engines, plus the
# tests, which are legitimate consumers of a diagnostic.
outside = src
for engine in ENGINES:
    i = outside.find("function " + engine)
    if i < 0:
        continue
    j = outside.index(") {", i) + 2
    depth, k = 0, j
    while k < len(outside):
        if outside[k] == "{":
            depth += 1
        elif outside[k] == "}":
            depth -= 1
            if depth == 0:
                break
        k += 1
    outside = outside[:i] + outside[k:]

import glob
import os
test_src = ""
for f in glob.glob(os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "tests", "**", "*.js"), recursive=True):
    test_src += open(f, encoding="utf-8").read()

# A field read inside the engine that set it is a working variable, not a dead
# output — an earlier version excluded the engine bodies from the search and so
# reported every one of those as orphaned.
engine_bodies = ""
for engine in ENGINES:
    i = src.find("function " + engine)
    if i < 0:
        continue
    j = src.index(") {", i) + 2
    depth, k = 0, j
    while k < len(src):
        if src[k] == "{":
            depth += 1
        elif src[k] == "}":
            depth -= 1
            if depth == 0:
                break
        k += 1
    engine_bodies += src[j:k]

orphans = []
for field, engines in sorted(set_fields.items()):
    pattern = r"\.\s*" + field + r"\b"
    if re.search(pattern, outside) or re.search(pattern, test_src):
        continue
    # read by its own engine, other than the assignment itself
    reads = len(re.findall(pattern, engine_bodies))
    writes = len(re.findall(r"out\." + field + r"\s*=(?![=])", engine_bodies))
    if reads > writes:
        continue
    orphans.append((field, sorted(engines)))

if orphans:
    print("FAIL engine result fields nothing reads:")
    for field, engines in orphans:
        print(f"    out.{field}  (set by {', '.join(engines)})")
    sys.exit(1)
print(f"OK   every engine result field has a reader ({len(set_fields)} checked)")

File: scripts/verify/livecheck.mjs
#!/usr/bin/env node
/* Ported from legacy/tools/livecheck.py — the highest-value port. No
 * off-the-shelf equivalent exists for this.
 *
 * Fields declared in a config table must be read by the app, not only by
 * tests. STABILITY_RULES carried `greenPerDay`/`amberPerDay` on eight
 * parameters that nothing in the app ever read — drift is graded elsewhere,
 * by CONSISTENCY_RULES. The only readers in the whole repository were the
 * tests. A stage-3 audit found alkalinity "graded steady... three times the
 * published limit", fixed it, and verified the fix by reading the same
 * unused constant: a fix and a test agreeing with each other about a number
 * the app never consults. deadcode.mjs cannot see this — the constant IS
 * read, just not by anything that runs.
 *
 * Change from the monolith: each table now lives in its own module, so "read
 * anywhere outside the table's own file" widens to "read anywhere in src/
 * outside the declaring file" — still excluding tests, which is the entire
 * point: a field read only by tests must still fail.
 */
import fs from 'node:fs';
import path from 'node:path';
import { ROOT, listSrcFiles, readAll, rel, matchBracket } from './util.mjs';

const TABLES = {
  STABILITY_RULES: 'src/lib/stability-engine.js',
  CONSISTENCY_RULES: 'src/lib/analytics/time-in-range.js',
  SAFE_BOUNDS: 'src/lib/findings.js',
  KIT_PRECISION: 'src/lib/findings.js',
  CORRECTION_MAX_RATE: 'src/lib/analytics/safe-rate.js',
  DOSE_ADVICE_RULES: 'src/lib/analytics/drift.js',
  DOSE_DRIFT_TRIGGER: 'src/lib/dosing/helpers.js',
};

const files = listSrcFiles();
const entries = readAll(files);
const problems = [];

for (const [table, relPath] of Object.entries(TABLES)) {
  const file = path.join(ROOT, relPath);
  const src = fs.readFileSync(file, 'utf8');
  const m = src.match(new RegExp(`(?:export )?const ${table}\\s*=\\s*\\{`));
  if (!m) { problems.push(`${table}: declaration not found in ${relPath} — table moved, re-point livecheck`); continue; }
  const open = m.index + m[0].length - 1;
  const close = matchBracket(src, open);
  const body = src.slice(open, close + 1);
  const fields = new Set([...body.matchAll(/[{,]\s*(\w+):/g)].map((f) => f[1]));

  // "Outside" = every application module minus the table's own declaration —
  // same file with the declaration text removed, plus every OTHER module.
  const outsideSameFile = src.slice(0, m.index) + src.slice(close + 1);
  const otherFiles = entries.filter((e) => e.path !== file).map((e) => e.src).join('\n');
  const outside = outsideSameFile + '\n' + otherFiles;

  for (const field of [...fields].sort()) {
    if (['why', 'label', 'unit', 'mode', 'displayPer'].includes(field)) continue; // descriptive, not consulted
    const readPattern = new RegExp(`\\.\\s*${field}\\b`);
    if (readPattern.test(outside) || outside.includes(`"${field}"`) || outside.includes(`'${field}'`)) continue;
    problems.push(`${table}.${field}  (${relPath})`);
  }
}

if (problems.length) {
  console.log('FAIL config fields the app never reads:');
  for (const p of problems) console.log(`    ${p}`);
  console.log('    (delete them, or wire them up — a field only tests read is worse than dead)');
  process.exit(1);
}
console.log(`OK   every config field is read by the app (${Object.keys(TABLES).length} tables checked)`);

// ---------------------------------------------------------------------------
// Fields an engine SETS on its result (`out.field = `) that nothing reads.
// `out.previous` was set by assessCalcium alone and read by nothing anywhere
// — a field one copy of three grew, that no consumer asked for.
const ENGINES = {
  assessAlkalinity: 'src/lib/dosing/alkalinity.js',
  assessCalcium: 'src/lib/dosing/calcium.js',
  assessMagnesium: 'src/lib/dosing/helpers.js',
};

function functionBody(src, name) {
  const i = src.indexOf('function ' + name);
  if (i < 0) return null;
  const openParen = src.indexOf('(', i);
  const closeParen = matchBracket(src, openParen);
  const openBrace = src.indexOf('{', closeParen);
  const closeBrace = matchBracket(src, openBrace);
  return src.slice(openBrace, closeBrace + 1);
}

const setFields = {};
const engineSources = {};
for (const [engine, relPath] of Object.entries(ENGINES)) {
  const src = fs.readFileSync(path.join(ROOT, relPath), 'utf8');
  engineSources[engine] = src;
  const body = functionBody(src, engine);
  if (!body) continue;
  for (const m of body.matchAll(/\bout\.(\w+)\s*=(?!=)/g)) {
    (setFields[m[1]] = setFields[m[1]] || new Set()).add(engine);
  }
}

const appCorpus = entries.map((e) => e.src).join('\n');
let testCorpus = '';
for (const dir of ['tests/legacy-port', 'src/test']) {
  const abs = path.join(ROOT, dir);
  if (!fs.existsSync(abs)) continue;
  const walk = (d) => {
    for (const ent of fs.readdirSync(d, { withFileTypes: true })) {
      const p = path.join(d, ent.name);
      if (ent.isDirectory()) walk(p);
      else if (/\.js$/.test(ent.name)) testCorpus += fs.readFileSync(p, 'utf8');
    }
  };
  walk(abs);
}

// Remove the engine bodies from the "outside" corpus so a field read inside
// the same engine that set it (a working variable) isn't mistaken for a
// consumer — but keep those bodies to check "read by its own engine" below.
let outsideApp = appCorpus;
for (const engine of Object.keys(ENGINES)) {
  const body = functionBody(engineSources[engine], engine);
  if (body) outsideApp = outsideApp.split(body).join('');
}

const orphans = [];
for (const [field, engines] of Object.entries(setFields).sort()) {
  const pattern = new RegExp(`\\.\\s*${field}\\b`);
  if (pattern.test(outsideApp) || pattern.test(testCorpus)) continue;
  let reads = 0, writes = 0;
  for (const engine of engines) {
    const body = functionBody(engineSources[engine], engine) || '';
    reads += (body.match(new RegExp(`\\.\\s*${field}\\b`, 'g')) || []).length;
    writes += (body.match(new RegExp(`out\\.${field}\\s*=(?!=)`, 'g')) || []).length;
  }
  if (reads > writes) continue;
  orphans.push(`out.${field}  (set by ${[...engines].sort().join(', ')})`);
}

if (orphans.length) {
  console.log('FAIL engine result fields nothing reads:');
  for (const o of orphans) console.log(`    ${o}`);
  process.exit(1);
}
console.log(`OK   every engine result field has a reader (${Object.keys(setFields).length} checked)`);
